api.proto

```diff
@@ -275,13 +275,13 @@ message Message {
     string              ackInbox        = 10; // NATS subject to publish acks to
     string              correlationId   = 11; // User-supplied value to correlate acks to publishes
     AckPolicy           ackPolicy       = 12; // Controls the behavior of acks
-    int64               lastKnownOffset = 13; // Provide last known offset on the partion before publish
+    int64               expectedOffset  = 13; // Provide expected offset on the partion after publish
```

Comment:
This shouldn't be needed on `Message`. The only way to receive a `Message` is if it was successfully published, which means `expectedOffset` will always equal `offset`.

Reply:
Make senses. In case of error it will be in the `AckError` already, otherwise we never need this in `Message`

Reply:
I suppose the message should be published on NATS topic anyway, whether with correct or incorrect offset.

It is in `messageProcessingLoop` of the partition leader that the verification is performed.

At the moment of reception at [`publishingLoop`](https://github.com/liftbridge-io/liftbridge/blob/603b7037427b98b7e65ce1293281e8ffe2c36d2e/server/api.go#L816), the server anyway needs to publish the message to NATS, then delegate the error handling on `AckError`

```golang
		msg, err := proto.MarshalPublish(&client.Message{
			Key:           req.Key,
			Value:         req.Value,
			Stream:        req.Stream,
			Subject:       subject,
			Headers:       req.Headers,
			AckInbox:      req.AckInbox,
			CorrelationId: req.CorrelationId,
			AckPolicy:     req.AckPolicy,
                        // Here, somehow the message should include the expected offset
		})

		if err := p.ncPublishes.Publish(subject, msg); err != nil {
			err = errors.Wrap(err, "failed to publish to NATS")
			p.logger.Errorf("api: Failed to publish async message: %v", err)
			p.sendPublishAsyncError(req.CorrelationId, &client.PublishAsyncError{
				Code:    client.PublishAsyncError_INTERNAL,
				Message: err.Error(),
			})
		}

```

Without including `expectedOffset` directly in the message, I guess it's hard to let the partition's leader knows what to expect

Reply:
Ah, yes sorry you're right. For some reason thought there was separate structs for "ingoing" and "outgoing" messages.

Reply:
That being said, we _could_ use the existing `offset` field on `Message` for the client to indicate the _expected_ offset and the server to set the _actual_ offset. Thoughts?

Reply:
That being said, we could use the existing offset field on Message for the client to indicate the _expected_ offset and the server to set the _actual_ offset. This field is currently ignored by the server. One issue is distinguishing between a `0` value and unset, but that issue exists even if it's a separate field.

Thoughts?

Reply:
It means that for now, the field number 1 ( `offset`) is not being used ? If that the case I think we can use it for the same purpose as the `expected` offset.

The only reason I added an extra `expectedOffset` is that I am not sure if reusing `offset` will break anything.

For 0 value, I am thinking if it is even make senses to consider 0 to be a valid expected offset ? As any non-null message will generate an expected offset bigger than 0.

So in general, the rule of server side would be:

```
if offset > 0:
    do optimistic offset control:
        calculate expected offset if the message is appended to log
        if incorrect offset then return error
        append log as normal
        return true
else append log as normal 
```

Reply:
> It means that for now, the field number 1 ( offset) is not being used ? If that the case I think we can use it for the same purpose as the expected offset.

It's only set by the server when sending messages to the client. For incoming messages to the server, it is ignored (because the server assigns the offset, not the client).

> For 0 value, I am thinking if it is even make senses to consider 0 to be a valid expected offset ? As any non-null message will generate an expected offset bigger than 0.

It's valid for the very first message in the log, which will have an offset of 0. This would preclude a client from doing optimistic concurrency control for the first message.

Reply:
At creation, the `newestOffset` is -1, so for the edge case that the very first message may increase expected offset by 1, which makes `expectedOffset` to be 0, indeed the optimistic concurrency control will not be performed.

To avoid that, I'm afraid choices are limited. We may have to add some kind of `NullableExpectedOffset` in here to differentiate between actual 0 and null value.

The first option is simple enough but without explicit notice to client side, users may encounter strange behavior for those edge cases.

The second option is rather complete, but I am wondering if it is a bit "overkill" ?

Reply:
Another option, as outlined in the [similar Kafka proposal](https://issues.apache.org/jira/browse/KAFKA-2260), is to use -1 as a sigil to indicate "next available offset" (current behavior) and put the OCC behavior behind a global or per-stream config to avoid breaking existing clients. I'm torn on what the right approach is.

Reply:
But in that case, if the client does:

1. Create a stream with OCC on
2. Do not set `offset` field, which makes it 0 by default

Then the server will actually consider 0 as a valid value.

I guess the main point is to let the server know which 0 is the "good one" and which 0 is the one to ignore.

Putting a OCC config flag per stream seems to be a better choice that using a customized `NullableExpectedOffset` , as it follows the same logics of configuration and does not create any "exception" (`NullableExpectedOffset` is nothing but an exception)

What do you think ?

Reply:
The config flag feels like the best option to me. I think clients should set the `offset` to -1 by default to indicate current behavior. Then the flag can be used to avoid breaking existing clients by making OCC opt-in on streams.

Reply:
As [EventStore](https://github.com/EventStore/EventStore) implements this functionality, here the [relevant bits from their documentation](https://developers.eventstore.com/server/5.0.8/http-api/writing-events.html#expected-version-header):

> There are special values you can use in the expected version header:
>
> -2 states that this write should never conflict and should always succeed.
> -1 states that the stream should not exist at the time of the writing (this write creates it).
> 0 states that the stream should exist but should be empty.

Reply:
I added in the latest commit :
- Optimistic Concurrency flag on stream creation config
- Remove `expectedOffset` in `Message`. `offset` field of `Message` will be used instead in the future